### PR TITLE
extended resource can cancelation support

### PR DESCRIPTION
**Pull request by xh3b4sd**

### CODE REVIEW

File: framework/context/reconciliationcanceledcontext/reconciliation_canceled_context.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
+// Package reconciliationcanceledcontext stores and accesses the canceled in
+// context.Context.
+package reconciliationcanceledcontext
```
Comment (xh3b4sd) [inline framework/context/reconciliationcanceledcontext/reconciliation_canceled_context.go]: This package is what `canceledcontext` was before. So `reconciliationcanceledcontext` is for canceling the complete reconciliation of the object. I am not sure about the naming. Naming is hard. Feel free to suggest something better.

File: framework/context/resourcecanceledcontext/resource_canceled_context.go [hunk 2]
```diff
@@ -1,6 +1,6 @@
-// Package canceledcontext stores and accesses the canceled in
+// Package resourcecanceledcontext stores and accesses the canceled in
 // context.Context.
-package canceledcontext
+package resourcecanceledcontext
```
Comment (xh3b4sd) [inline framework/context/resourcecanceledcontext/resource_canceled_context.go]: Package `resourcecanceledcontext` is to cancel the resource only. Following resources are still reconciled even though a resource got canceled.

File: framework/framework.go [hunk 3]
```diff
@@ -349,9 +351,13 @@ func ProcessDelete(ctx context.Context, obj interface{}, resources []Resource) e
...
 		var currentState interface{}
 		{
-			if canceledcontext.IsCanceled(ctx) {
+			if reconciliationcanceledcontext.IsCanceled(ctx) {
 				return nil
 			}
+			if resourcecanceledcontext.IsCanceled(ctx) {
+				ctx = resourcecanceledcontext.NewContext(ctx, make(chan struct{}))
+				continue
+			}
```
Comment (xh3b4sd) [inline framework/framework.go]: When we cancel the complete loop we just return. When we cancel a single resource we want to continue with the next resource. We also have to make sure the context is initialized with a new channel we can close again. This is a tiny but important implementation detail. It is not really nice, but it works.

File: framework/framework_test.go [hunk 4]
```diff
@@ -80,13 +81,13 @@ func Test_Framework_ResourceCallOrder(t *testing.T) {
...
-		// the expected order until it gets canceled.
+		// the expected order until the reconciliation gets canceled.
 		{
 			ProcessMethod: ProcessCreate,
-			Ctx:           canceledcontext.NewContext(context.Background(), make(chan struct{})),
+			Ctx:           reconciliationcanceledcontext.NewContext(context.Background(), make(chan struct{})),
 			Resources: []Resource{
 				&testResource{
-					CancelingStep: "GetCurrentState",
+					ReconciliationCanceledStep: "GetCurrentState",
```
Comment (xh3b4sd) [inline framework/framework_test.go]: I renamed things for the existing tests.

File: framework/framework_test.go [hunk 5]
```diff
@@ -98,14 +99,56 @@ func Test_Framework_ResourceCallOrder(t *testing.T) {
...
+					ResourceCanceledStep: "GetCurrentState",
+				},
+			},
+			ExpectedOrders: [][]string{
+				{
+					"GetCurrentState",
+				},
+			},
+			ErrorMatcher: nil,
+		},
```
Comment (xh3b4sd) [inline framework/framework_test.go]: I added new tests for the resource cancelation.

File: framework/framework_test.go [hunk 6]
```diff
@@ -144,16 +187,76 @@ func Test_Framework_ResourceCallOrder(t *testing.T) {
...
+					"GetCurrentState",
+					"GetDesiredState",
+					"NewUpdatePatch",
+					"ApplyCreatePatch",
+					"ApplyDeletePatch",
+					"ApplyUpdatePatch",
+				},
+			},
+			ErrorMatcher: nil,
+		},
```
Comment (xh3b4sd) [inline framework/framework_test.go]: This case shows quite well how the cancelation of a single resource works.

File: framework/context/reconciliationcanceledcontext/reconciliation_canceled_context.go [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+		}
+	}
+
+	return false
+}
+
+// SetCanceled is a safe way to signal cancelation.
+func SetCanceled(ctx context.Context) {
+	canceled, canceledExists := FromContext(ctx)
+	if canceledExists && !IsCanceled(ctx) {
```
Comment (JosephSalisbury) [inline framework/context/reconciliationcanceledcontext/reconciliation_canceled_context.go]: Do we need to check `canceledExists` here? Don't we check it in `IsCanceled`?

File: framework/context/reconciliationcanceledcontext/reconciliation_canceled_context.go (hunk 1 above)
Comment (kopiczko) [reply framework/context/reconciliationcanceledcontext/reconciliation_canceled_context.go]: What is `canceledcontext` now?

Comment (xh3b4sd) [reply framework/context/reconciliationcanceledcontext/reconciliation_canceled_context.go]: > This package is what `canceledcontext` was before.

There is no `canceledcontext` anymore. It is now `reconciliationcanceledcontext`.

File: framework/context/reconciliationcanceledcontext/reconciliation_canceled_context.go (hunk 7 above)
Comment (xh3b4sd) [reply framework/context/reconciliationcanceledcontext/reconciliation_canceled_context.go]: I get what you mean. You are probably right. Note that this here is not new but just copy/paste from the package we already had before.

File: framework/context/reconciliationcanceledcontext/reconciliation_canceled_context.go (hunk 1 above)
Comment (kopiczko) [reply framework/context/reconciliationcanceledcontext/reconciliation_canceled_context.go]: I understand from the comment and code below.

**Pull request merged**